### PR TITLE
Allow the style property of block.json to be an array and add support for object-based block styles

### PR DESCRIPTION
**Pull request by adamziel**
## What problem does this PR solve?

In https://github.com/WordPress/gutenberg/pull/34180 we moved block CSS to `blockJson.supports.__experimentalStyle`. This PR moves it under `blockJson.style` at the root level. The goal is to only have a single location where styles are defined.

Block.json **before** this PR:

```js
{
	"supports":  {
		"__experimentalStyle": {
			"border": {			
				"color": {
					"text": "#fff",
					"background": "#32373c"
				}
			}
		}
	},
	"style":  "wp-block-button"
}
```

Block.json **after** this PR:

```js
{
	"style":  [ "wp-block-button", {
		"border": {			
			"color": {
				"text": "#fff",
				"background": "#32373c"
			}
		}
	}]
}
```

This PR is an alternative to https://github.com/WordPress/gutenberg/pull/41656 

## Other considerations

This is a breaking change – any existing code that assumes `style` to be a single item or an array of strings will throw a notice at best.

## Test plan

1. Apply this PR to your local wordpress-develop
2. Run `npm run build`
3. Apply https://github.com/WordPress/gutenberg/pull/41873 to your local Gutenberg installation
4. Build Gutenberg
5. Switch themes to TwentyTwentyTwo
6. Create a new page
7. Insert a button, give it some text
8. Save it, view it on the frontend
9. Confirm the buttons have a large padding at the bottom

cc @getdave @scruffian @draganescu 

Trac ticket: https://core.trac.wordpress.org/ticket/56094

### CODE REVIEW

**Comment by scruffian** [comment]
> This is a breaking change – any existing code that assumes style to be a single item or an array of strings will throw a notice at best.

This seems like a blocker. Can we change it to allow for strings as well?

File: src/wp-includes/block-editor.php
Comment (scruffian) [inline src/wp-includes/block-editor.php]: Same questions as on the GB PR - why only use the first one?

File: src/wp-includes/blocks.php
Comment (scruffian) [inline src/wp-includes/blocks.php]: This probably needs a better name.

**Comment by adamziel** [comment]
> This seems like a blocker. Can we change it to allow for strings as well?

@scruffian oh it can be a string. I'm saying it can be either of:

* A string
* An array of strings
* A mixed array of strings and arrays

Therefore, any code that assumes it's just a string or string[] will break.

File: src/wp-includes/block-editor.php
Comment (adamziel) [reply src/wp-includes/block-editor.php]: Without this PR, `_wp_multiple_block_styles` replaces the `style` property from an array to the first element of that array. With this PR, the `style` property remains an array but we still want to enqueue/register only its first element. That's what happens here.

File: src/wp-includes/blocks.php
Comment (adamziel) [reply src/wp-includes/blocks.php]: Agreed! Any suggestions?

File: src/wp-includes/block-editor.php
Comment (scruffian) [reply src/wp-includes/block-editor.php]: Right, I wonder if we can make it a bit smarter. Maybe we can add all the strings and filter out all the arrays?

File: src/wp-includes/blocks.php
Comment (scruffian) [reply src/wp-includes/blocks.php]: What about `register_single_style_handle_for_block`? I don't think we need the `_` prefix.

File: src/wp-includes/block-editor.php
Comment (adamziel) [reply src/wp-includes/block-editor.php]: If we filter out all the arrays we won't be able to use the information from these arrays. As a result, [`WP_Theme_JSON_Resolver_Gutenberg`](https://github.com/WordPress/gutenberg/pull/41873/files#diff-b960a542dcc41109b5d5daf17726ff7ecc22a07f7c6c87dbfdea6cf44dd1e3b0) won't be able get them rendered. That's the pickle we're in: Updating `style` to accept arrays means we have to deal with arrays now.

File: src/wp-includes/blocks.php
Comment (adamziel) [reply src/wp-includes/blocks.php]: 👍

File: src/wp-includes/block-editor.php
Comment (scruffian) [reply src/wp-includes/block-editor.php]: When I say filter them out I mean simply don't set them on $style_handles.

Like this:

```
if ( is_array( $block_type->style ) ) {
	foreach( $block_type->style as $single_style) {
		if ( ! is_array( $single_style) ) {
			$style_handles[] = $single_style;
		}
	}
}
```

Comment (adamziel) [reply src/wp-includes/block-editor.php]: Gotcha! I just did that.

Comment (scruffian) [reply src/wp-includes/block-editor.php]: Nice. Do we need to do the same in the Gutenberg PR?

**Comment by adamziel** [comment]
@scruffian I applied all the feedback and refreshed the related Gutenberg PR. This one is ready for another review!

**Comment by gziolo** [comment]
@adamziel, thank you so much for looking into adding native support for multiple `style` styles per block type. It's all going in the right direction. It will require a bit more work to propagate changes to other places:

- `WP_Block_Type` class:
  https://github.com/WordPress/wordpress-develop/blob/a5b555aac3b95d368efd2c69c057d23d836acc86/src/wp-includes/class-wp-block-type.php#L199-L205
- `WP_REST_Block_Types_Controller` class:
  https://github.com/WordPress/wordpress-develop/blob/a5b555aac3b95d368efd2c69c057d23d836acc86/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L544-L550

This will include also existing unit tests for both classes. New functionality requires also unit test coverage.

I don't think we consume `style` from the REST API endpoint in Gutenberg, but we should double-check to be on the safe side. 

We should also extend the existing documentation for `block.json` to include the newly supported format of styles:
https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#style

What is the reason this patch covers only `style` but not `editorStyle`?

File: src/wp-includes/blocks.php
Comment (gziolo) [inline src/wp-includes/blocks.php]: So this changes the behavior of both the `editorStyle` and `style`, but we have only changed handling for `style` in the rest of the codebase.

@aristath, can we get rid of this `_wp_multiple_block_styles` callback if we add native handling for an array in `WP_Block_Type`?

Comment (aristath) [reply src/wp-includes/blocks.php]: > @aristath, can we get rid of this `_wp_multiple_block_styles` callback if we add native handling for an array in `WP_Block_Type`?

If `WP_Block_Type` can natively handle arrays then I don't see why not... Yes, we could remove this is the implementation gets merged in `WP_Block_Type` 👍

Comment (gziolo) [reply src/wp-includes/blocks.php]: One thing we could do is to keep support for the string when defining `style` or `editoStyle` when calling `register_block_type`, but convert it behind the scenes to an array when assigning to `WP_Block_Type` properties.

Comment (aristath) [reply src/wp-includes/blocks.php]: Would make perfect sense, yes 😄

**Comment by gziolo** [comment]
We will have to put an empty `_wp_multiple_block_styles` function with the deprecation message, example:

https://github.com/WordPress/wordpress-develop/blob/0abc2b2a5bbc6737a7c31d28456037be91a0bfa7/src/wp-includes/deprecated.php#L4305-L4313

File: src/wp-includes/class-wp-block-type.php
Comment (gziolo) [inline src/wp-includes/class-wp-block-type.php]: You can put that function directly in the `WP_Block_Type` class and call from the method that integrates `register_block_type_args` filter.

File: src/wp-includes/blocks.php
Comment (adamziel) [reply src/wp-includes/blocks.php]: Done!

File: src/wp-includes/class-wp-block-type.php
Comment (gziolo) [inline src/wp-includes/class-wp-block-type.php]: It's no longer a filter.

By convention, every method should have `@since` documented.

File: src/wp-includes/class-wp-block-type.php
Comment (gziolo) [inline src/wp-includes/class-wp-block-type.php:263]: We should also document with `@since` that since 6.1.0 an array format is supported. We should also expand the description because it's no longer a style handle only.

**Comment by gziolo** [comment]
It looks like we have everything covered so it needs a final review and a round of testing with and without the Gutenberg plugin.

**Comment by adamziel** [comment]
I think we're good now! GitHub CI seems to be down, though, so might be worth waiting until it's up again before merging.

**Comment by adamziel** [comment]
All the tests passed! Would you please help get this PR approved? CC @gziolo @aristath @scruffian @getdave

**Review by scruffian — APPROVED** [review]
LGTM

**Comment by adamziel** [comment]
@gziolo do you think you could merge this one?

**Review by aristath — APPROVED** [review]

**Comment by adamziel** [comment]
cc @mtias for soundness check – does this change seem reasonable to you?

**Comment by gziolo** [comment]
The example shared in the description looks nice in my opinion as it consolidates several available options. In the future, we could also use the same API borrowed from `theme.json` for style variations:

```json
{
	"styles": [
		{
			"name": "default",
			"label": "Default",
			"isDefault": true
		},
		{
			"name": "other",
			"label": "Other",
			"settings": {
				"border": {			
					"color": {
						"text": "#000",
						"background": "#ccccc"
					}
				}
			}
		 }
	],
	"style":  [ "wp-block-button", {
		"border": {			
			"color": {
				"text": "#fff",
				"background": "#32373c"
			}
		}
	} ]
}
```

**Comment by gziolo** [comment]
Let's land https://github.com/WordPress/gutenberg/pull/41873 in the Gutenberg plugin first and give it a try before we include it in WordPress core. Code-wise everything is ready, and it had enough reviewers. We mostly need to validate that the extended API for `style` in `block.json` is friendly for developers – it's mostly the question of whether it's intuitive and straightforward to document.

**Comment by adamziel** [comment]
I'm closing this in favor of https://github.com/WordPress/wordpress-develop/pull/3108

**Pull request closed without merging**
